packages/react-reconciler/src/ReactFiberNewContext.js

```diff
@@ -264,46 +264,41 @@ export function readContext<T>(
   context: ReactContext<T>,
   observedBits: void | number | boolean,
 ): T {
-  if (typeof observedBits !== 'number') {
-    if (observedBits === false) {
-      // Do not observe updates
-      observedBits = 0;
+  if (lastContextWithAllBitsObserved === context) {
+    // Nothing to do. We already observe everything in this context.
+  } else if (observedBits === false || observedBits === 0) {
+    // Do not observe any updates.
+  } else {
+    let resolvedObservedBits; // Avoid deopting on observable arguments or heterogeneous types.
+    if (
+      typeof observedBits !== 'number' ||
+      observedBits === maxSigned31BitInt
+    ) {
+      // Observe all updates.
+      lastContextWithAllBitsObserved = ((context: any): ReactContext<mixed>);
+      resolvedObservedBits = maxSigned31BitInt;
     } else {
-      // Observe all updates
-      observedBits = maxSigned31BitInt;
+      resolvedObservedBits = observedBits;
     }
-  }
 
-  if (lastContext === null) {
-    invariant(
-      currentlyRenderingFiber !== null,
-      'Context.unstable_read(): Context can only be read while React is ' +
-        'rendering, e.g. inside the render method or getDerivedStateFromProps.',
-    );
-    // This is the first dependency in the list
-    currentlyRenderingFiber.firstContextDependency = lastContextDependency = {
+    let contextItem = {
       context: ((context: any): ReactContext<mixed>),
-      observedBits,
+      observedBits: resolvedObservedBits,
       next: null,
     };
-    lastContext = context;
-  } else {
-    // `lastContextDependency` is always non-null if `lastContext is.
-    const lastDependency: ContextDependency<T> = (lastContextDependency: any);
-    if (lastContext === context) {
-      // Fast path. The previous context has the same type. We can reuse
-      // the same node.
-      lastDependency.observedBits |= observedBits;
```

Comment:
We no longer have this optimization. To add it back, we would have to track an additional field globally. (We need to save on these since it adds cost to each thread in a multi-threaded renderer). We also have to check that additional field when we add it - i.e. add another branch here. For that to be worth it, this pattern have to be somewhat common to justify the optimization. I'd argue that it's not.